[CI] StatelessResizeIT testSplitIndex failing

**Build Scans:**
- [elasticsearch-intake #40175 / part2](https://gradle-enterprise.elastic.co/s/vnhyznhdijary)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:stateless:internalClusterTest" --tests "org.elasticsearch.xpack.stateless.StatelessResizeIT.testSplitIndex" -Dtests.seed=481581E45E1BCF5F -Dtests.locale=pt-MZ -Dtests.timezone=Africa/Asmara -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.stateless.StatelessResizeIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testSplitIndex'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: pending-stateless_commit_7-T0tGdG-pQ6GU08B1kphitQ
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testSplitIndex
- [main] 1 consecutive failures in step part2
- [main] 1 consecutive failures in pipeline elasticsearch-intake

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -603,9 +603,6 @@ tests:
 - class: org.elasticsearch.reindex.management.ListTasksRelocationIT
   method: testListTasksShowsOriginalIdentityAfterRelocation
   issue: https://github.com/elastic/elasticsearch/issues/149776
-- class: org.elasticsearch.xpack.stateless.StatelessResizeIT
-  method: testSplitIndex
-  issue: https://github.com/elastic/elasticsearch/issues/149777
 - class: org.elasticsearch.xpack.dlm.frozen.DLMFrozenTransitionWithSecurityIT
   method: testCanTransitionIndex
   issue: https://github.com/elastic/elasticsearch/issues/149770

--- a/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/AbstractStatelessPluginIntegTestCase.java
+++ b/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/AbstractStatelessPluginIntegTestCase.java
@@ -1197,6 +1197,14 @@ public abstract class AbstractStatelessPluginIntegTestCase extends ESIntegTestCa
         return new PrimaryTermAndGeneration(indexShard.getOperationPrimaryTerm(), ((IndexEngine) engineOrNull).getCurrentGeneration());
     }
 
+    /**
+     * Call this when the blobs you want listed have been fully uploaded, e.g., after a flush,
+     * or call this in a loop (e.g., assertBusy) to check that some blobs have been deleted.
+     *
+     * If blobs are being uploaded, some non-conforming filenames may not be listed, e.g., if they
+     * have partial names like the temporary "pending-" prefixed blobs that {@link FsBlobContainer}
+     * uses to do atomic uploads.
+     */
     protected static Set<PrimaryTermAndGeneration> listBlobsTermAndGenerations(ShardId shardId) throws Exception {
         Set<PrimaryTermAndGeneration> set = new HashSet<>();
         var objectStoreService = getObjectStoreService(internalCluster().getRandomNodeName());
@@ -1204,11 +1212,11 @@ public abstract class AbstractStatelessPluginIntegTestCase extends ESIntegTestCa
         for (var entry : indexBlobContainer.children(operationPurpose).entrySet()) {
             var primaryTerm = Long.parseLong(entry.getKey());
             Set<String> statelessCompoundCommits = entry.getValue().listBlobs(operationPurpose).keySet();
-            statelessCompoundCommits.forEach(
-                filename -> set.add(
-                    new PrimaryTermAndGeneration(primaryTerm, StatelessCompoundCommit.parseGenerationFromBlobName(filename))
-                )
-            );
+            statelessCompoundCommits.forEach(filename -> {
+                if (StatelessCompoundCommit.startsWithBlobPrefix(filename)) {
+                    set.add(new PrimaryTermAndGeneration(primaryTerm, StatelessCompoundCommit.parseGenerationFromBlobName(filename)));
+                }
+            });
         }
         return set;
     }

--- a/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/StatelessResizeIT.java
+++ b/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/StatelessResizeIT.java
@@ -214,6 +214,7 @@ public class StatelessResizeIT extends AbstractStatelessPluginIntegTestCase {
         ensureGreen(shrunkIndex);
 
         assertHitCount(prepareSearch(shrunkIndex).setSize(0).setQuery(QueryBuilders.matchAllQuery()), numDocs);
+        flush(shrunkIndex);
         for (int s = 0; s < shrunkShards; s++) {
             final ShardId shardId = findIndexShard(resolveIndex(shrunkIndex), s).shardId();
             assertThat(listBlobsTermAndGenerations(shardId), hasSize(greaterThan(0)));
@@ -353,14 +354,12 @@ public class StatelessResizeIT extends AbstractStatelessPluginIntegTestCase {
         if (numNewDocs > 0) {
             indexDocsAndRefresh(targetIndex, numNewDocs);
         }
-        if (randomBoolean()) {
-            flush(targetIndex);
-        }
 
         // Verify all documents are present in the target index
         assertHitCount(prepareSearch(targetIndex).setSize(0).setQuery(QueryBuilders.matchAllQuery()), numDocs + numNewDocs);
 
         // Verify that each target shard has its own blob
+        flush(targetIndex);
         for (int s = 0; s < targetShards; s++) {
             final ShardId targetShardId = findIndexShard(resolveIndex(targetIndex), s).shardId();
             assertThat(listBlobsTermAndGenerations(targetShardId), hasSize(greaterThan(0)));

